htdocs/product/stock/class/mouvementstock.class.php

```diff
@@ -484,7 +484,7 @@ public function _create($user, $fk_product, $entrepot_id, $qty, $type, $price =
 			{
 				// Note: PMP is calculated on stock input only (type of movement = 0 or 3). If type == 0 or 3, qty should be > 0.
 				// Note: Price should always be >0 or 0. PMP should be always >0 (calculated on input)
-				if (($type == 0 || $type == 3) && $price > 0)
+				if (($type == 0 || $type == 3) && $price >= 0)
```

Comment:
If price is zero (or free), it should not have effect on the AWP (by definition, the AWP is the average price when product is not free, doing this will update the AWP lowering the AWP). 
If you really need this, because this is not common, you should introduce a test on a hidden option (For example if $conf->global->STOCK_UPDATE_AWP_EVEN_WHEN_ENTRY_PRICE_IS_NULL) and do your rule only when this option is set.

Reply:
Hi, 
thank you for your answer. Do you have a link to an official definition of AWP / PMP ? (I can't get anything precise).
My accountant (expert-comptable) seems to be familiar with the consideration of free products, and also in very specific cases to negatives AWP because of returned product with a higher price. I don't want to mean I want negatives AWP in my stock, but this is a good marker to detect errors.
Thank you for your time. 
If I want to update the formula in my case, your way seems good. To keep Dolibarr unmodified, I must add this option in a specific personal module.
(Sorry for my questions/suggestions, I'm not a pro developper, we are preparing a big instance of Dolibarr with multi-company for my shop network - you came for a formation a year ago. We are working with administrators and developers. I will add those elements to the planning.)
Best Regards